modules/ipp/attributevalues.go

```diff
@@ -0,0 +1,288 @@
+package ipp
+
+import (
+	"encoding/binary"
+	"bytes"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type IPPTime struct {
+	Year uint16
+	Month byte
+	Day byte
+	Hour byte
+	Minutes byte
+	Seconds byte
+	Deciseconds byte
+	DirectionFromUTC byte
+	HoursFromUTC byte
+	MinutesFromUTC byte
+}
+
+type Resolution struct {
+	XFeed int32 `json:"cross-feed-direction"`
+	Feed int32 `json:"feed-direction"`
+	Unit string `json:"unit"`
+
+}
+
+type RangeOfInteger struct {
+	Min int32 `json:"min"`
+	Max int32 `json:"max"`
+}
+
+type StringWithLanguage struct {
+	Lang string `json:"language"`
+	String string `json:"stringWithoutLanguage"`
+}
+
+// TODO: Move buf := bytes.NewBuffer declaration to the top of this function, rather than repeating EVERYWHERE
+func getParse(b byte) func(*AttrValue) {
+	switch {
+// Out-of-Band Values
+//	// unsupported
+//	case b == 0x10:
+//
+//	// unknown
+//	case b == 0x12:
+//
+//	// no-value
+//	case b == 0x13:
+
+
+// Integer Types
+	// Integer
+	case b == 0x21:
+		return func(val *AttrValue) {
+			buf := bytes.NewBuffer(val.Bytes)
+			var i int32
+			if err := binary.Read(buf, binary.BigEndian, &i); err != nil {
+				log.WithFields(log.Fields{
+					"error": err,
+					"data": val.Bytes,
+				}).Debug("Failed to interpret data with error.")
+				return
+			}
+			val.Integer = &i
+		}
+	// Boolean
+	case b == 0x22:
+		return func(val *AttrValue) {
+			if len(val.Bytes) == 1 {
+				var truth bool
+				switch val.Bytes[0] {
+				case 0x00:
+					truth = false
+				case 0x01:
+					truth = true
+				}
+				val.Boolean = &truth
+			}
+		}
+	// Enum
+	case b == 0x23:
+		return func(val *AttrValue) {
+			// TODO: Implement
+		}
+
+// octetString Types
+	// octetString
+	case b == 0x30:
+		return func(val *AttrValue) {
+			// TODO: Figure out the best way to implement
+			val.OctetString = val.Bytes
+		}
+	// dateTime
+	case b == 0x31:
+		return func(val *AttrValue) {
+			buf := bytes.NewBuffer(val.Bytes)
+			t := &IPPTime{}
+			if err := binary.Read(buf, binary.BigEndian, t); err != nil {
+				log.WithFields(log.Fields{
+					"error": err,
+					"data": val.Bytes,
+				}).Debug("Failed to interpret data with error.")
+				return
+			}
+			// Seconds East of UTC
+			// + -> East
+			// - -> West
+			secondsEast := int(t.DirectionFromUTC) * 60 * (int(t.MinutesFromUTC) + (60 * int(t.HoursFromUTC)))
```

Comment:
Shouldn't the last term be multiplied by another 60?

Reply:
I think the answer is two-fold:
1) No.
2) My parentheses make this really unclear.
That last hours term is only multiplied by 60, then added to the minutes term. Then that sum is multiplied by 60 again to get seconds.

Reply:
Duh, right --  I missed the nested parens.